fix(api): Unify code to create/edit Dataset

wido opened the pull request:
Just as done with c330668a9f550c8d82b7cf393ba71ef9544fd02d for ZVOL this commit updates the datapath for the Dataset so that both the GUI and API use the same.

Signed-off-by: Wido den Hollander <wido@widodh.nl>

wido commented:
@william-gr It took me a bit of time as Django and the FreeNAS code is pretty new to me, but here is a first stab of it.

I've tested as much as possible, but I wanted to open the PR to see if I took the right route. Could you review it?

william-gr commented on gui/middleware/notifier.py:
```diff
@@ -3315,6 +3315,7 @@ def zfs_set_option(self, name, item, value, recursive=False):
         """
         name = str(name)
         item = str(item)
+        log.error('name: %s item: %s value: %s' % (name, item, value))
```
Debug leftover

wido replied:
Removing that one!

william-gr commented on gui/storage/forms.py:
```diff
@@ -130,6 +130,12 @@ def _clean_zfssize_fields(form, attrs, prefix):
+            continue
+        except ValueError:
+            pass
+
```
That check looks wrong to me. cdata[formfield] may have letters, e.g. "10M" which will then be converted.

wido replied:
No, it will throw a ValueError if there is 10M.

When merely a int is provided we can continue to the next item.

william-gr replied:
Well, then it will skip all the rest, thats some another bug, can you leave it off and I'll see whats up with that?

wido replied:
It will test all of the rest indeed. Otherwise the other code will throw an error that you need to specify a human format like 10M, 10G, 1T, etc.

By testing if it's a pure int you know you have a valid value and can continue.

william-gr commented on gui/storage/forms.py:
```diff
@@ -1350,13 +1288,34 @@ def clean_dataset_name(self):
+        for prop in ['org.freenas:description', 'compression', 'atime', 'dedup',
+                     'share_type', 'recordsize', 'casesensitivity']:
+            if prop == 'org.freenas:description':
+                value = self.cleaned_data.get('dataset_org.freenas:description')
```
Looks like it should be .get('dataset_comments') ?

william-gr commented on gui/storage/forms.py:
```diff
@@ -1274,74 +1279,7 @@ class ZFSDataset(Form):
-                'Contact <a href="mailto:truenas-support@ixsystems.com?subject'
-                '=ZFS Deduplication Activation">TrueNAS Support</a> for '
-                'assistance.</span><br />'
-            )
```
You have removed all these from the Create forms which seems wrong. It will no longer report the inherit value on create time nor show dedup warning.

william-gr commented on gui/storage/forms.py:
```diff
@@ -1367,26 +1326,129 @@ def clean_dataset_recordsize(self):
+        errors = dict()
+
+        for item, value in props.items():
+            log.error('fs: %s item: %s value: %s' % (name, item, value))
```
Another debug leftover

wido commented:
@william-gr: I have pushed a additional commit which addresses your concerns.

I am now left with one problem which I am not able to resolve.

When *UPDATING* a dataset through the API you now have to supply every propery. Eg, you have to submit atime, compression, quota, etc. Otherwise it will complain that it's a required value.

Now, if I set a field to not required other things break.

What would the best route be in this case?

william-gr commented:
Awesome, thanks for the changes!

I think you have to get the values on the API update method and feed the form with that, overwriting with user provided values:

cur_data = GETDATASETOPTIONS
data = cur_data.update(data)

But I don't want to demand too much from you. If you prefer I can take this from now on.

wido commented:
@william-gr: Thanks! If you would? That would be great.

My main intention was the ability to set the quota when creating a Dataset and updating it afterwards through the API.

Reason is that I want to be able to create users and datasets API-driven. Users will log in through FTP/SSH and need a quota on their homedir.

Would be great if this could make it into 11.1

william-gr commented:
@wido sure thing! I can have it in by 11.1.

Do you mind if I create a new branch, cherry-pick your continue and continue from there?

william-gr commented:
And here is the ticket to track it https://bugs.freenas.org/issues/24930
